src/protocols/rdp/rdp.c

```diff
@@ -227,10 +227,53 @@ static BOOL rdp_freerdp_authenticate(freerdp* instance, char** username,
 
     rdpContext* context = instance->context;
     guac_client* client = ((rdp_freerdp_context*) context)->client;
-
-    /* Warn if connection is likely to fail due to lack of credentials */
-    guac_client_log(client, GUAC_LOG_INFO,
-            "Authentication requested but username or password not given");
+    guac_rdp_client* rdp_client = (guac_rdp_client*) client->data;
+    guac_rdp_settings* settings = rdp_client->settings;
+    char* params[4] = {};
+    int i = 0;
+    
+    if (settings->username == NULL || strcmp(settings->username, "") == 0) {
+        params[i] = "username";
+        rdp_client->rdp_credential_flags |= GUAC_RDP_CRED_FLAG_USERNAME;
+        i++;
+    }
+    
+    if (settings->password == NULL || strcmp(settings->password, "") == 0) {
+        params[i] = "password";
+        rdp_client->rdp_credential_flags |= GUAC_RDP_CRED_FLAG_PASSWORD;
+        i++;
+    }
+    
+    if (settings->domain == NULL || strcmp(settings->domain, "") == 0) {
+        params[i] = "domain";
+        rdp_client->rdp_credential_flags |= GUAC_RDP_CRED_FLAG_DOMAIN;
+        i++;
+    }
+    
+    /* NULL-terminate the array. */
+    params[i] = NULL;
+    
+    if (i > 0) {
+        /* Lock the client thread. */
+        pthread_mutex_lock(&(rdp_client->rdp_credential_lock));
+        
+        /* Send require params and flush socket. */
+        guac_protocol_send_required(client->socket, (const char**) params);
+        guac_socket_flush(client->socket);
```

Comment:
Okay, I have taken a wild stab at implementing support for this on the server-side.  It compiles, now, but I haven't made the client-side changes that will be necessary to make it work.

I've no doubt that it needs some polishing, so let me know what needs to be changed.

Reply:
Just a quick update to this - I'm sure you can gather most of it from looking at the code, but the idea I have is that:
*Dduring the handshake process, another instruction will be addded - `protocol` - which the client will use to communicate its protocol level to the server
* guacd will store that version, and, if it doesn't receive a version, will have a default one.  My thought is that the default will be VERSION_1_1_0, since that's the latest version that doesn't require any special instructions client-side.
* The rest of the process will proceed per usual - client and server will negotiate the latest version common to them, and then both sides will store that information for the duration of the session.